File: go/v1/s3url.go
package v1

import (
	"errors"
	"strings"
)

func newS3UrlFromS3Bucket(v S3Bucket) (string, error) {
	if v.Bucket == "" || v.Prefix == "" {
		return "", errors.New("invalid S3 URL: must specify both Bucket and Prefix")
	}

	return "s3://" + v.Bucket + "/" + v.Prefix, nil
}

func newS3UrlFromS3Object(v S3Object) (string, error) {
	if v.Bucket == "" || v.ObjectKey == "" {
		return "", errors.New("invalid S3 URL: must specify both Bucket and ObjectKey")
	}

	return "s3://" + v.Bucket + "/" + v.ObjectKey, nil
}

func splitS3Url(url string) (string, string, error) {
	tokens := strings.Split(url, "//")
	if len(tokens) != 2 {
		return "", "", errors.New("invalid S3 URL: must be of format s3://bucket-name/key-name")
	}

	tokens = strings.Split(tokens[1], "/")
	if len(tokens) < 2 {
		return "", "", errors.New("invalid S3 URL: must be of format s3://bucket-name/key-name")
	}

	return tokens[0], strings.Join(tokens[1:], "/"), nil
}

File: python/awsutils/s3_url.py
def split_s3_url(url: str) -> dict:
    tokens = url.split("//")
    if len(tokens) < 2:
        raise ValueError("invalid S3 URL: must be of format s3://bucket-name/key-name")

    tokens = tokens[1].split("/")
    if len(tokens) < 2:
        raise ValueError("invalid S3 URL: must be of format s3://bucket-name/key-name")

    return {'bucket': tokens[0], 'objectKey': '/'.join(tokens[1:])}
